File: app/src/main/java/com/example/macarus0/walkiewalkie/view/SelectOwnerActivity.java
package com.example.macarus0.walkiewalkie.view;

import android.app.Activity;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Intent;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;

import com.example.macarus0.walkiewalkie.R;
import com.example.macarus0.walkiewalkie.viewmodel.WalkieViewModel;

import butterknife.BindView;
import butterknife.ButterKnife;

public class SelectOwnerActivity extends AppCompatActivity implements OwnerListAdapter.OwnerClickHandler{

    public static final String OWNER_ID = "owner_id";

    WalkieViewModel mViewModel;
    @BindView(R.id.select_contact_toolbar)
    Toolbar mToolbar;
    @BindView(R.id.select_items_list)
    RecyclerView mItemsRecyclerView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_select_contact);
        ButterKnife.bind(this);
        mViewModel = ViewModelProviders.of(this).get(WalkieViewModel.class);
        setSupportActionBar(mToolbar);
        // add back arrow to toolbar
        if (getSupportActionBar() != null){
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            getSupportActionBar().setDisplayShowHomeEnabled(true);
            getSupportActionBar().setTitle(R.string.select_owner);
        }

        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this);
        mItemsRecyclerView.setLayoutManager(linearLayoutManager);

        OwnerListAdapter ownerListAdapter = new OwnerListAdapter();
        ownerListAdapter.setOwnerClickHandler(this::ownerClick);
        ownerListAdapter.setShowDogs(false);
        mItemsRecyclerView.setAdapter(ownerListAdapter);
        mViewModel.getAllAvailableOwners().observe(this, ownerListAdapter::setOwners);
    }

    @Override
    public void ownerClick(long id) {
        Intent intent = new Intent();
        intent.putExtra(OWNER_ID, id);
        setResult(Activity.RESULT_OK, intent);
        finish();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // handle arrow click here
        if (item.getItemId() == android.R.id.home) {
            finish();
        }

        return super.onOptionsItemSelected(item);
    }


}

File: app/src/main/java/com/example/macarus0/walkiewalkie/view/SelectOwnerActivity.kt
package com.example.macarus0.walkiewalkie.view

import android.app.Activity
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.appcompat.widget.Toolbar
import android.view.MenuItem

import com.example.macarus0.walkiewalkie.R
import com.example.macarus0.walkiewalkie.viewmodel.WalkieViewModel

import butterknife.BindView
import butterknife.ButterKnife
import com.example.macarus0.walkiewalkie.data.Owner

class SelectOwnerActivity : AppCompatActivity() {

    private lateinit var mViewModel: WalkieViewModel
    @BindView(R.id.select_contact_toolbar)
    @JvmField  var mToolbar: Toolbar? = null
    @BindView(R.id.select_items_list)
    @JvmField  var mItemsRecyclerView: RecyclerView? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_select_contact)
        ButterKnife.bind(this)
        mViewModel = ViewModelProviders.of(this).get(WalkieViewModel::class.java)
        setSupportActionBar(mToolbar)
        // add back arrow to toolbar
        if (supportActionBar != null) {
            supportActionBar!!.setDisplayHomeAsUpEnabled(true)
            supportActionBar!!.setDisplayShowHomeEnabled(true)
            supportActionBar!!.setTitle(R.string.select_owner)
        }

        val intent = intent
        val dogId = intent.getLongExtra(DOG_ID, -1)

        val linearLayoutManager = LinearLayoutManager(this)
        mItemsRecyclerView!!.layoutManager = linearLayoutManager

        val ownerListAdapter = OwnerListAdapter()
        ownerListAdapter.ownerClickHandler = {it ->  this.ownerClick(it) }
        ownerListAdapter.setShowDogs(false)
        mItemsRecyclerView!!.adapter = ownerListAdapter
        mViewModel.getAllAvailableOwners(dogId).observe(this, Observer<List<Owner>> { ownerListAdapter.setOwners(requireNotNull(it)) })
    }

    fun ownerClick(id: Long) {
        val intent = Intent()
        intent.putExtra(OWNER_ID, id)
        setResult(Activity.RESULT_OK, intent)
        finishAfterTransition()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // handle arrow click here
        if (item.itemId == android.R.id.home) {
            finishAfterTransition()
        }

        return super.onOptionsItemSelected(item)
    }

    companion object {

        const val OWNER_ID = "owner_id"
        const val DOG_ID = "dog_id"
    }


}
